### PR TITLE
XWIKI-17648: NestedMentionsIT#basic is flickering

### PR DESCRIPTION
**Pull request by manuelleduc**
https://jira.xwiki.org/browse/XWIKI-17648

More elements tested, comments mentions tested too.
Retry on flickering elements introduced.

### CODE REVIEW

File: xwiki-platform-core/xwiki-platform-mentions/xwiki-platform-mentions-test/xwiki-platform-mentions-test-docker/src/test/it/org/xwiki/mentions/test/ui/MentionsIT.java [hunk 1]
```diff
@@ -67,6 +74,20 @@
         void run() throws Exception;
     }
 
+    @AfterEach
+    void tearDown(TestUtils setup, TestReference reference) throws Exception
+    {
+        // clear all notifications between each test.
+        runAsUser(setup, U2_USERNAME, USERS_PWD, () -> {
```
Comment (surli) [inline xwiki-platform-core/xwiki-platform-mentions/xwiki-platform-mentions-test/xwiki-platform-mentions-test-docker/src/test/it/org/xwiki/mentions/test/ui/MentionsIT.java:81]: Why not deleting the users between test instead?

File: xwiki-platform-core/xwiki-platform-mentions/xwiki-platform-mentions-test/xwiki-platform-mentions-test-docker/src/test/it/org/xwiki/mentions/test/ui/MentionsIT.java [hunk 2]
```diff
@@ -141,4 +229,31 @@ private void runAsSuperAdmin(TestUtils setup, RunnableErr actions) throws Except
         setup.loginAsSuperAdmin();
         actions.run();
     }
+
+    private <T> T retry(int times, Callable<T> callable) throws Exception
```
Comment (surli) [inline xwiki-platform-core/xwiki-platform-mentions/xwiki-platform-mentions-test/xwiki-platform-mentions-test-docker/src/test/it/org/xwiki/mentions/test/ui/MentionsIT.java:233]: That's not good, you shouldn't handle the test retry inside your own test. Test retry should be done at the execution level: we already have what's needed for it, it's just that we don't enable it usually for timeout reason.

File: xwiki-platform-core/xwiki-platform-mentions/xwiki-platform-mentions-test/xwiki-platform-mentions-test-pageobjects/src/main/java/org/xwiki/mentions/test/po/MentionNotificationPage.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    private static final String EMITTER_SELECTOR =
+        "tr td span.notification-event-user";
+
+    private static final String SUMMARY_SELECTOR = "../following-sibling::tr/td/blockquote";
+
+    /**
+     * Opens a notification group.
+     * @param notificationNumber The notification index.
+     */
+    public void openGroup(int notificationNumber)
```
Comment (surli) [inline xwiki-platform-core/xwiki-platform-mentions/xwiki-platform-mentions-test/xwiki-platform-mentions-test-pageobjects/src/main/java/org/xwiki/mentions/test/po/MentionNotificationPage.java:53]: It doesn't look related to Mention to me, sounds like a general need for Notifications test instead.

Comment (surli) [reply xwiki-platform-core/xwiki-platform-mentions/xwiki-platform-mentions-test/xwiki-platform-mentions-test-pageobjects/src/main/java/org/xwiki/mentions/test/po/MentionNotificationPage.java:53]: It's same for the other methods: they are all general to notifications apparently. Maybe you could introduce a GroupedNotificationElement PO for all of them, but shouldn't be tied to mention.

File: xwiki-platform-core/xwiki-platform-mentions/xwiki-platform-mentions-test/xwiki-platform-mentions-test-docker/src/test/it/org/xwiki/mentions/test/ui/MentionsIT.java (hunk 1 above)
Comment (manuelleduc) [reply xwiki-platform-core/xwiki-platform-mentions/xwiki-platform-mentions-test/xwiki-platform-mentions-test-docker/src/test/it/org/xwiki/mentions/test/ui/MentionsIT.java:81]: Just clearing the notifications is a lighter operation, but overall the result would be the same.

Comment (vmassol) [reply xwiki-platform-core/xwiki-platform-mentions/xwiki-platform-mentions-test/xwiki-platform-mentions-test-docker/src/test/it/org/xwiki/mentions/test/ui/MentionsIT.java:81]: Actually tearDown is bad practice in general. You should have a setup instead. And a single one since it's a scenario.

File: xwiki-platform-core/xwiki-platform-mentions/xwiki-platform-mentions-test/xwiki-platform-mentions-test-docker/pom.xml [hunk 4]
```diff
@@ -60,9 +60,8 @@
     <!-- page objects -->
     <dependency>
       <groupId>org.xwiki.platform</groupId>
-      <artifactId>xwiki-platform-notifications-test-pageobjects</artifactId>
+      <artifactId>xwiki-platform-mentions-test-pageobjects</artifactId>
       <version>${project.version}</version>
-      <scope>test</scope>
```
[the remark this answers is not in the document]
Comment (manuelleduc) [reply xwiki-platform-core/xwiki-platform-mentions/xwiki-platform-mentions-test/xwiki-platform-mentions-test-docker/pom.xml:65]: Fixed

File: xwiki-platform-core/xwiki-platform-notifications/xwiki-platform-notifications-test/xwiki-platform-notifications-test-pageobjects/src/main/java/org/xwiki/platform/notifications/test/po/GroupedNotificationElementPage.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        "tr td span.notification-event-user";
+
+    /**
+     * Opens a notification group.
+     * @param notificationNumber The notification index.
+     */
+    public void openGroup(int notificationNumber)
+    {
+        WebElement notifications =
+            getDriver().findElementsByCssSelector(NOTIFICATION_EVENT_SELECTOR).get(notificationNumber);
```
[the remark this answers is not in the document]
Comment (manuelleduc) [reply xwiki-platform-core/xwiki-platform-notifications/xwiki-platform-notifications-test/xwiki-platform-notifications-test-pageobjects/src/main/java/org/xwiki/platform/notifications/test/po/GroupedNotificationElementPage.java:53]: I've added the closed parent with an id to the css selector to limit the search scope of the query.

[the remark this answers is not in the document]
Comment (surli) [reply xwiki-platform-core/xwiki-platform-notifications/xwiki-platform-notifications-test/xwiki-platform-notifications-test-pageobjects/src/main/java/org/xwiki/platform/notifications/test/po/GroupedNotificationElementPage.java:53]: What if you want to use this page object for a notif seen on the dashboard?

File: xwiki-platform-core/xwiki-platform-notifications/xwiki-platform-notifications-test/xwiki-platform-notifications-test-pageobjects/src/main/java/org/xwiki/platform/notifications/test/po/GroupedNotificationElementPage.java [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+        WebElement emitter = notifications.findElements(cssSelector(EMITTER_SELECTOR)).get(groupIndex);
+        return emitter.findElement(By.tagName("a")).getText();
+    }
+
+    /**
+     *
+     * @param driver XWIki web driver.
+     * @return The root element of the notification tray.
+     */
+    public static WebElement getRootElementNotificationTray(XWikiWebDriver driver)
```
Comment (surli) [inline xwiki-platform-core/xwiki-platform-notifications/xwiki-platform-notifications-test/xwiki-platform-notifications-test-pageobjects/src/main/java/org/xwiki/platform/notifications/test/po/GroupedNotificationElementPage.java:104]: I wouldn't have done it like this: we already have a PO for the Notification Tray so it should be its responsibility to create a GroupedNotificationElementPage with the right parent IMO.

File: xwiki-platform-core/xwiki-platform-mentions/xwiki-platform-mentions-test/xwiki-platform-mentions-test-docker/src/test/it/org/xwiki/mentions/test/ui/MentionsIT.java [hunk 7]
```diff
@@ -99,8 +107,72 @@ void basic(TestUtils setup, TestReference reference) throws Exception
...
+     *     <li>Superadmin creates U1 and U2.</li>
+     *     <li>U1 adds a mention to U2.</li>
+     *     <li>U2 verify that she has received a notification.</li>
+     * </ul>
+     *
+     * @param setup The test setup.
+     * @param reference The test page reference.
+     * @throws Exception In case of error.
+     * @since 12.7
+     * @since 12.6.2
```
Comment (surli) [inline xwiki-platform-core/xwiki-platform-mentions/xwiki-platform-mentions-test/xwiki-platform-mentions-test-docker/src/test/it/org/xwiki/mentions/test/ui/MentionsIT.java:145]: We usually don't put since on test. And anyway those needs to be updated.

File: xwiki-platform-core/xwiki-platform-mentions/xwiki-platform-mentions-test/xwiki-platform-mentions-test-pageobjects/src/main/java/org/xwiki/mentions/test/po/MentionNotificationPage.java [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+
+import static org.openqa.selenium.By.cssSelector;
+import static org.openqa.selenium.By.xpath;
+
+/**
+ * Mentions notifications page object.
+ *
+ * @version $Id$
+ * @since 12.7
+ * @since 12.6.2
```
Comment (surli) [inline xwiki-platform-core/xwiki-platform-mentions/xwiki-platform-mentions-test/xwiki-platform-mentions-test-pageobjects/src/main/java/org/xwiki/mentions/test/po/MentionNotificationPage.java:35]: Here too.

Comment (surli) [reply xwiki-platform-core/xwiki-platform-mentions/xwiki-platform-mentions-test/xwiki-platform-mentions-test-pageobjects/src/main/java/org/xwiki/mentions/test/po/MentionNotificationPage.java:35]: Basically the only since you should put here for now is 12.8RC1: we're not sure yet to backport this to branches.

**Comment by surli** [comment]
Thanks!
